src/metadata_map.rs

```diff
@@ -0,0 +1,1716 @@
+/// more than once if it has more than one associated value.
+#[derive(Debug)]
+pub struct Iter<'a, T: 'a> {
+    inner: http::header::Iter<'a, T>,
+}
+
+/// A drain iterator of all values associated with a single metadata key.
+#[derive(Debug)]
+pub struct ValueDrain<'a, T: 'a> {
+    inner: http::header::ValueDrain<'a, T>,
+}
+
+/// A drain iterator for `MetadataMap`.
+#[derive(Debug)]
+pub struct Drain<'a, T: 'a> {
+    inner: http::header::Drain<'a, T>,
+}
+
+/// An iterator over `MetadataMap` keys.
+///
+/// Each header name is yielded only once, even if it has more than one
+/// associated value.
+#[derive(Debug)]
+pub struct Keys<'a, T: 'a> {
+    inner: http::header::Keys<'a, T>,
+}
+
+/// `MetadataMap` value iterator.
+///
+/// Each value contained in the `MetadataMap` will be yielded.
+#[derive(Debug)]
+pub struct Values<'a, T: 'a> {
+    inner: http::header::Values<'a, T>,
+}
+
+/// An iterator of all values associated with a single metadata key.
+#[derive(Debug)]
+pub struct ValueIter<'a, T: 'a> {
+    inner: http::header::ValueIter<'a, T>,
+}
+
+/// An iterator of all values associated with a single metadata key.
+#[derive(Debug)]
+pub struct ValueIterMut<'a, T: 'a> {
+    inner: http::header::ValueIterMut<'a, T>,
+}
+
+/// A view to all values stored in a single entry.
+///
+/// This struct is returned by `MetadataMap::get_all`.
+#[derive(Debug)]
+pub struct GetAll<'a> {
+    inner: http::header::GetAll<'a, MetadataValue>
+}
+
+/// A view into a single location in a `MetadataMap`, which may be vacant or occupied.
+#[derive(Debug)]
+pub enum Entry<'a, T: 'a> {
+    /// An occupied entry
+    Occupied(OccupiedEntry<'a, T>),
+
+    /// A vacant entry
+    Vacant(VacantEntry<'a, T>),
+}
+
+/// A view into a single empty location in a `MetadataMap`.
+///
+/// This struct is returned as part of the `Entry` enum.
+#[derive(Debug)]
+pub struct VacantEntry<'a, T: 'a> {
+    inner: http::header::VacantEntry<'a, T>,
+}
+
+/// A view into a single occupied location in a `MetadataMap`.
+///
+/// This struct is returned as part of the `Entry` enum.
+#[derive(Debug)]
+pub struct OccupiedEntry<'a, T: 'a> {
+    inner: http::header::OccupiedEntry<'a, T>,
+}
+
+// ===== impl MetadataMap =====
+
+impl MetadataMap {
+    /// Create an empty `MetadataMap`.
+    ///
+    /// The map will be created without any capacity. This function will not
+    /// allocate.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use tower_grpc::metadata::*;
+    /// let map = MetadataMap::new();
+    ///
+    /// assert!(map.is_empty());
+    /// assert_eq!(0, map.capacity());
+    /// ```
+    pub fn new() -> Self {
+        MetadataMap::with_capacity(0)
+    }
+
+    /// Convert an HTTP HeaderMap to a MetadataMap
+    pub fn from_headers(mut headers: http::HeaderMap) -> Self {
+        let mut map = Self::with_capacity(headers.len());
+        for (name, values) in headers.drain() {
```

Comment:
`http::HeaderMap` is equal to `http::HeaderMap<http::HeaderValue>`. `MetadataMap` however has a `http::HeaderMap<MetadataValue>`, so when constructing a `MetadataMap` the values need to be wrapped individually.

`MetadataMap` can't just wrap a `http::HeaderMap<http::HeaderValue>` because there are some important methods that return references to those values, which would be unimplementable then (or would require a separate `MetadataValueRef` type that wrapped `HeaderName&`).

(There is a similar issue for methods that return references to map keys but I worked around that by returning `&str`s instead, which is basically what `HeaderName` is anyways after construction and you don't need to quickly compare them with standard HTTP header names.)

At least this code just moves all the headers, so runtime is proportional to header count but not size.

Reply:
We could still do it, though it would require an unsafe transmute. stdlib relies on this being valid in a number of places. For example, `Path::new` where `Path(OsStr)`:

```rust
    pub fn new<S: AsRef<OsStr> + ?Sized>(s: &S) -> &Path {
        unsafe { &*(s.as_ref() as *const OsStr as *const Path) }
    }
```

I would use this strategy personally. I think a zero-cost `MetadataMap` wrapper is critical here, otherwise switching is not be worth it.

Reply:
Cool idea! Done.